iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp

```diff
@@ -284,53 +182,71 @@ static linalg::LinalgLoopDistributionOptions convPoolDistributionOptions = {
     {linalg::DistributionMethod::Cyclic, linalg::DistributionMethod::Cyclic,
      linalg::DistributionMethod::Cyclic}};
 
-/// Pattern for tiling convolution and pooling operations. Currently is just a
-/// way to not tile when the operation has padding.
+/// Pattern for tiling convolution and pooling operations.
 template <typename OpTy>
 struct TileConvPoolPattern : public linalg::LinalgTilingPattern<OpTy> {
   using Base = linalg::LinalgTilingPattern<OpTy>;
   TileConvPoolPattern(MLIRContext *context, linalg::LinalgTilingOptions options,
-                      ArrayRef<int64_t> tileSizes,
-                      ArrayRef<int64_t> workgroupSize,
+                      const LaunchConfig &launchConfig,
                       PatternBenefit benefit = 1)
       : Base(context,
              options.setDistributionOptions(convPoolDistributionOptions),
              linalg::LinalgMarker(
                  ArrayRef<Identifier>(),
                  Identifier::get(getWorkgroupMarker(), context)),
              benefit),
-        tileSizes(tileSizes.begin(), tileSizes.end()),
-        workgroupSize(workgroupSize.begin(), workgroupSize.end()) {}
+        launchConfig(launchConfig) {}
 
   LogicalResult matchAndRewrite(Operation *op,
                                 PatternRewriter &rewriter) const override {
     if (hasPadding(cast<OpTy>(op))) return failure();
     FuncOp funcOp = op->getParentOfType<FuncOp>();
     if (!funcOp || failed(Base::matchAndRewrite(op, rewriter)) ||
-        failed(updateWorkGroupSize(funcOp, this->workgroupSize)))
+        failed(updateWorkGroupSize(funcOp, launchConfig.getWorkgroupSize())))
       return failure();
     funcOp.removeAttr(getNumWorkgroupsFnAttrName());
     return success();
   }
 
-  SmallVector<int64_t, 3> tileSizes;
-  SmallVector<int64_t, 3> workgroupSize;
+  const LaunchConfig &launchConfig;
 };
 
+/// Populate patterns for first-level tiling.
+static void populateFirstLevelTilingPatterns(
+    MLIRContext *context, const LaunchConfig &launchConfig,
+    OwningRewritePatternList &patterns) {
+  // Function to compute first level tiling values.
+  std::function<SmallVector<Value, 4>(OpBuilder &, Operation *)>
```

Comment:
might just use `auto` here? I don't think we care about the detailed function type here?